pkg/graveler/ref/commit_ordered_iterator.go

```diff
@@ -55,23 +72,43 @@ func (iter *OrderedCommitIterator) maybeFetch() {
 	if len(iter.buf) > 0 {
 		return
 	}
+	psql := sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+	q := psql.Select("id", "committer", "message", "creation_date", "meta_range_id", "parents", "metadata", "version").
+		From("graveler_commits").
+		Where(sq.Eq{"repository_id": iter.repositoryID})
 
-	var offsetCondition string
 	if iter.state == iteratorStateInit {
-		offsetCondition = iteratorOffsetCondition(true)
 		iter.state = iteratorStateQuerying
+		q = q.Where(sq.GtOrEq{"id": iter.offset})
 	} else {
-		offsetCondition = iteratorOffsetCondition(false)
+		q = q.Where(sq.Gt{"id": iter.offset})
 	}
 
 	var buf []*commitRecord
-	err := iter.db.Select(iter.ctx, &buf, `
-			SELECT id, committer, message, creation_date, meta_range_id, parents, metadata, version
-			FROM graveler_commits
-			WHERE repository_id = $1
-			AND id `+offsetCondition+` $2
-			ORDER BY id ASC
-			LIMIT $3`, iter.repositoryID, iter.offset, iter.prefetchSize)
+
+	if iter.onlyAncestryLeaves {
+		notExistsCondition := psql.Select("*").
+			Prefix("NOT EXISTS (").Suffix(")").
+			From("graveler_commits c2").
+			Where("c2.repository_id=graveler_commits.repository_id").
+			Where(sq.Or{
+				sq.And{
+					sq.Lt{"c2.version": graveler.CommitVersionParentSwitch},
+					sq.Expr("c2.parents[array_length(c2.parents, 1)]=graveler_commits.id")},
+				sq.And{
+					sq.GtOrEq{"c2.version": graveler.CommitVersionParentSwitch},
+					sq.Expr("c2.parents[1]=graveler_commits.id")},
+			})
```

Comment:
It _might_ be possible to create indices on `graveler_commits` for second and last parents:

```sql
CREATE INDEX graveler_commits_second_parent_idx ((parents[2]));
CREATE INDEX graveler_commits_last_parent_idx ((parents[array_length(parents, 1)]));
```
Note however that, especially for the second index, you will have to be _very_ careful to use an expression that the query planner can recognise as being equal to the index.

Reply:
Thanks Ariel! I'm looking into it